### ISSUE TITLE
Choosing models with distinct covariates

### ISSUE THREAD

**Issue by gcasamat** [microsoft/EconML#380]
Hi, 
I wonder if there is a way to compare the performance of causal models that use different set of covariates. 
I understand that you developed methods for comparing algorithms (score of second stage model on a validation set) and I would like to know if these methods are also valid for choosing which set of covariates to include in the model.
One concern is the fact that adding variables may induce missing values and thus change the set of observations used for estimation. 
Thank you.

**Comment by vsyrgkanis** [comment]
The scoring code, i.e. the RScorer expects that all models take the same features as input to “effect”.

but you can simulate what you want as follows:
Create a BaseCateEstimator child class that at init takes as input any Cate estimator and a subset of indices of features.
At fit time passes calls the fit of the wrapped cate estimator with X being the subset of features (and maybe the rest go to controls W)
At const_marginal_effect time calls the const_marginal_effect of the wrapped estimator on the subset of X

this way you created a wrapper of any cate estimator which coding wise takes all of X but actually fits the final model only on the subset.

tou can then pass all such cate estimators each instantiatrd with a different set of subset to the Rscorer, after you’ve trained them on a training set.

when you train make sure that eavh estimator takes as input the subset of training data for which all the X features used for training are not NaN (or impute missing values)

at test time use validation data for the Rscorer that contain all the features (or impute missing values in the same way as you did at training).

the imputation variant of the above might actually be preferable

**Comment by gcasamat** [comment]
Thank you for your reply ! 
Just a precision: when I wrote 'covariates', I meant variables belonging to the set W, not X (all the models I want to compare have the same set X). Does it make a difference with respect to the solution you propose?

**Comment by vsyrgkanis** [comment]
You should include all W. Unless you think that some W is an instrument based on domain knowledge.

but the set of W should not be chosen via cross validation and scoring. You should use all W and only exclude one if and only if you think that this W does not have a direct effect on Y or on T based on domain knowledge. Also if you have such domain knowledge then uou should exclude it from all cate estimators

**Comment by gcasamat** [comment]
ok, but suppose that I want to include lags of the outcome and treatment variables in W. What does it mean to include all W? How many lags should I include?

**Comment by vsyrgkanis** [comment]
Good question. I would go on the over side. Choose as many lags as you believe could “playsibly” have a direct effect on both the outcome and the treatment.

regularization (e.g. lasso or forest) in the first stage models would adaptively throw away variables that dont have predivtive value

**Comment by gcasamat** [comment]
Thank you, this is a clear recommendation. 
I understand that, in theory, we should provide with all the available information in W. 
In practice, however, I find that adding more lags may lead to "strange" results (e.g., positive price elasticities). Perhaps this is due to the fact that the introduction of additional lags induce missing values. Imputation of missing values may then help, as you suggest in a related issue.
